File: ACP_comm_test/ACP/acpcomms/java/Streamer.java
package com.usfiae.acp.commlib.acpcomms.java;

import java.net.*;
import java.util.Map;
import java.util.function.Consumer;


public class Streamer extends ACPComms {
    private DatagramSocket socket;
    private InetAddress address;
    private int port;
    private Thread listenerThread;
    private volatile boolean listening;
    private Consumer<DatagramPacket> packetHandler;
    private int bufferSize;

    public Streamer() {
        super();
    }

    @Override
    public void configure(Map<String, Object> configuration) throws Exception {
        this.config.putAll(configuration);

        String host = (String) config.getOrDefault("host", "localhost");
        this.port = (int) config.getOrDefault("port", 9999);
        this.address = InetAddress.getByName(host);
        this.bufferSize = (int) config.getOrDefault("bufferSize", 8192);

        Integer localPort = (Integer) config.get("localPort");
        if (localPort != null && socket == null) {
            socket = new DatagramSocket(localPort);
        }
    }


    @Override
    public void connect() throws Exception {
        if (connected) {
            throw new IllegalStateException("Already connected");
        }

        if (socket == null) {
            Integer localPort = (Integer) config.get("localPort");
            if (localPort != null) {
                socket = new DatagramSocket(localPort);
            } else {
                socket = new DatagramSocket();
            }
        }

        Integer timeout = (Integer) config.get("timeout");
        if (timeout != null) {
            socket.setSoTimeout(timeout);
        }

        connected = true;
    }


    @Override
    public void disconnect() throws Exception {
        if (!connected) {
            return;
        }

        stopListener();

        if (socket != null && !socket.isClosed()) {
            socket.close();
        }

        connected = false;
    }


    public void sendData(byte[] data) throws Exception {
        if (!connected) {
            throw new IllegalStateException("Not connected");
        }

        DatagramPacket packet = new DatagramPacket(data, data.length, address, port);
        socket.send(packet);
    }


    public void sendData(String data) throws Exception {
        sendData(data.getBytes());
    }


    public DatagramPacket receiveData() throws Exception {
        if (!connected) {
            throw new IllegalStateException("Not connected");
        }

        byte[] buffer = new byte[bufferSize];
        DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
        socket.receive(packet);
        return packet;
    }


    public void setPacketHandler(Consumer<DatagramPacket> handler) {
        this.packetHandler = handler;
    }


    public void startListener() throws Exception {
        if (!connected) {
            throw new IllegalStateException("Not connected");
        }
        if (listening) {
            throw new IllegalStateException("Listener already running");
        }

        listening = true;
        listenerThread = new Thread(() -> {
            while (listening && !Thread.currentThread().isInterrupted()) {
                try {
                    byte[] buffer = new byte[bufferSize];
                    DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                    socket.receive(packet);

                    if (packetHandler != null) {
                        packetHandler.accept(packet);
                    }
                } catch (SocketTimeoutException e) {
                    // Normal timeout, continue listening
                } catch (Exception e) {
                    if (listening) {
                        System.err.println("Error receiving packet: " + e.getMessage());
                    }
                }
            }
        });
        listenerThread.start();
    }


    public void stopListener() {
        listening = false;
        if (listenerThread != null) {
            listenerThread.interrupt();
            try {
                listenerThread.join(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }


    public InetAddress getAddress() {
        return address;
    } // Are these redundent?


    public int getPort() {
        return port;
    }
}

File: ACP_comm_test/ACP/acpcomms/python3/streamer.py
"""
Streamer - UDP-based streaming implementation
Uses Python standard library (socket module)
"""
import socket
import threading
from typing import Dict, Any, Callable, Optional, Tuple
from .acp_comms import ACPComms


class DatagramPacket:
    """
    Wrapper class to mimic Java's DatagramPacket
    Created to maintain interface compatibility
    """
    def __init__(self, data: bytes, address: Tuple[str, int]):
        self.data = data
        self.address = address
        self.length = len(data)
    
    def getData(self) -> bytes:
        """Get packet data"""
        return self.data
    
    def getAddress(self) -> Tuple[str, int]:
        """Get source address as (host, port) tuple"""
        return self.address
    
    def getLength(self) -> int:
        """Get data length"""
        return self.length


class Streamer(ACPComms):
    """UDP-based streaming implementation"""
    
    def __init__(self):
        super().__init__()
        self.socket: Optional[socket.socket] = None
        self.address: str = ""
        self.port: int = 0
        self.listenerThread: Optional[threading.Thread] = None
        self.listening: bool = False
        self.packetHandler: Optional[Callable[[DatagramPacket], None]] = None
        self.bufferSize: int = 8192
    
    def configure(self, configuration: Dict[str, Any]) -> None:
        """
        Configure the streamer
        
        Args:
            configuration: Dictionary containing:
                - host: Target host (default: "localhost")
                - port: Target port (default: 9999)
                - bufferSize: Receive buffer size (default: 8192)
                - localPort: Local port to bind to (optional)
                - timeout: Socket timeout in milliseconds (optional)
        """
        self.config.update(configuration)
        
        host = self.config.get("host", "localhost")
        self.port = self.config.get("port", 9999)
        self.address = host
        self.bufferSize = self.config.get("bufferSize", 8192)
        
        local_port = self.config.get("localPort")
        if local_port is not None and self.socket is None:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.socket.bind(("", local_port))
    
    def connect(self) -> None:
        """Initialize UDP socket"""
        if self.connected:
            raise RuntimeError("Already connected")
        
        if self.socket is None:
            local_port = self.config.get("localPort")
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            if local_port is not None:
                self.socket.bind(("", local_port))
        
        timeout = self.config.get("timeout")
        if timeout is not None:
            # Convert milliseconds to seconds for Python socket
            self.socket.settimeout(timeout / 1000.0)
        
        self.connected = True
    
    def disconnect(self) -> None:
        """Close UDP socket"""
        if not self.connected:
            return
        
        self.stopListener()
        
        if self.socket is not None:
            self.socket.close()
        
        self.connected = False
    
    def sendData(self, data: bytes) -> None:
        """
        Send data via UDP
        
        Args:
            data: Bytes to send
        """
        if not self.connected:
            raise RuntimeError("Not connected")
        
        self.socket.sendto(data, (self.address, self.port))
    
    def sendDataString(self, data: str) -> None:
        """
        Send string data via UDP
        
        Args:
            data: String to send
        """
        self.sendData(data.encode())
    
    def receiveData(self) -> DatagramPacket:
        """
        Receive data via UDP
        
        Returns:
            DatagramPacket: Received packet with data and address
        """
        if not self.connected:
            raise RuntimeError("Not connected")
        
        data, addr = self.socket.recvfrom(self.bufferSize)
        return DatagramPacket(data, addr)
    
    def setPacketHandler(self, handler: Callable[[DatagramPacket], None]) -> None:
        """
        Set the packet handler callback
        
        Args:
            handler: Callable that accepts DatagramPacket
        """
        self.packetHandler = handler
    
    def startListener(self) -> None:
        """Start listening for packets in a background thread"""
        if not self.connected:
            raise RuntimeError("Not connected")
        if self.listening:
            raise RuntimeError("Listener already running")
        
        self.listening = True
        
        def listener_loop():
            while self.listening:
                try:
                    data, addr = self.socket.recvfrom(self.bufferSize)
                    packet = DatagramPacket(data, addr)
                    
                    if self.packetHandler is not None:
                        self.packetHandler(packet)
                except socket.timeout:
                    # Normal timeout, continue listening
                    continue
                except Exception as e:
                    if self.listening:
                        print(f"Error receiving packet: {e}", flush=True)
        
        self.listenerThread = threading.Thread(target=listener_loop, daemon=True)
        self.listenerThread.start()
    
    def stopListener(self) -> None:
        """Stop the listener thread"""
        self.listening = False
        if self.listenerThread is not None:
            self.listenerThread.join(timeout=1.0)
    
    def getAddress(self) -> str:
        """
        Get target address
        
        Returns:
            str: Target host address
        """
        return self.address  # Are these redundant?
    
    def getPort(self) -> int:
        """
        Get target port
        
        Returns:
            int: Target port number
        """
        return self.port